lib/srv/db/postgres/engine.go

```diff
@@ -0,0 +1,454 @@
+	// messages b/w server and client e.g. to get client's password.
+	conn, err := pgconn.ConnectConfig(ctx, connectConfig)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	// Hijacked connection exposes some internal connection data, such as
+	// parameters we'll need to relay back to the client (e.g. database
+	// server version).
+	hijackedConn, err := conn.Hijack()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	frontend := pgproto3.NewFrontend(pgproto3.NewChunkReader(hijackedConn.Conn), hijackedConn.Conn)
+	return frontend, hijackedConn, nil
+}
+
+// makeClientReady indicates to the Postgres client (such as psql) that the
+// server is ready to accept messages from it.
+func (e *Engine) makeClientReady(client *pgproto3.Backend, hijackedConn *pgconn.HijackedConn) error {
+	// AuthenticationOk indicates that the authentication was successful.
+	e.Log.Debug("Sending AuthenticationOk.")
+	if err := client.Send(&pgproto3.AuthenticationOk{}); err != nil {
+		return trace.Wrap(err)
+	}
+	// BackendKeyData provides secret-key data that the frontend must save
+	// if it wants to be able to issue cancel requests later.
+	e.Log.Debugf("Sending BackendKeyData: PID=%v.", hijackedConn.PID)
+	if err := client.Send(&pgproto3.BackendKeyData{ProcessID: hijackedConn.PID, SecretKey: hijackedConn.SecretKey}); err != nil {
+		return trace.Wrap(err)
+	}
+	// ParameterStatuses contains parameters reported by the server such as
+	// server version, relay them back to the client.
+	e.Log.Debugf("Sending ParameterStatuses: %v.", hijackedConn.ParameterStatuses)
+	for k, v := range hijackedConn.ParameterStatuses {
+		if err := client.Send(&pgproto3.ParameterStatus{Name: k, Value: v}); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	// ReadyForQuery indicates that the start-up is completed and the
+	// frontend can now issue commands.
+	e.Log.Debug("Sending ReadyForQuery")
+	if err := client.Send(&pgproto3.ReadyForQuery{}); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// receiveFromClient receives messages from the provided backend (which
+// in turn receives them from psql or other client) and relays them to
+// the frontend connected to the database instance.
+func (e *Engine) receiveFromClient(client *pgproto3.Backend, server *pgproto3.Frontend, clientErrCh chan<- error, sessionCtx *session.Context) {
+	log := e.Log.WithField("from", "client")
+	defer log.Debug("Stop receiving from client.")
+	for {
+		message, err := client.Receive()
+		if err != nil {
+			log.WithError(err).Errorf("Failed to receive message from client.")
+			clientErrCh <- err
+			return
+		}
+		log.Debugf("Received client message: %#v.", message)
+		switch msg := message.(type) {
+		case *pgproto3.Query:
+			err := e.OnQuery(*sessionCtx, msg.String)
+			if err != nil {
+				log.WithError(err).Error("Failed to emit audit event.")
+			}
+		case *pgproto3.Terminate:
+			clientErrCh <- nil
+			return
+		}
+		err = server.Send(message)
```

Comment:
Is there any reason to not forward the `Terminate` message to the server?

Reply:
Yeah, it will be handled by "pgconn" - it will send 'X' (termination message) to the server and handle other wind-down logic when Close() method is called on it.